compiler/compiler.go

```diff
@@ -1,210 +1,136 @@
+		pp, err := readProto(file, importPaths, done)
+		if err != nil {
+			return nil, err
+		}
+		protos = append(protos, pp...)
 	}
+	return protos, nil
 }
 
-var scalars = map[parser.Scalar]pb.FieldDescriptorProto_Type{
-	parser.Double:   pb.FieldDescriptorProto_TYPE_DOUBLE,
-	parser.Float:    pb.FieldDescriptorProto_TYPE_FLOAT,
-	parser.Int32:    pb.FieldDescriptorProto_TYPE_INT32,
-	parser.Int64:    pb.FieldDescriptorProto_TYPE_INT64,
-	parser.Uint32:   pb.FieldDescriptorProto_TYPE_UINT32,
-	parser.Uint64:   pb.FieldDescriptorProto_TYPE_UINT64,
-	parser.Sint32:   pb.FieldDescriptorProto_TYPE_SINT32,
-	parser.Sint64:   pb.FieldDescriptorProto_TYPE_SINT64,
-	parser.Fixed32:  pb.FieldDescriptorProto_TYPE_FIXED32,
-	parser.Fixed64:  pb.FieldDescriptorProto_TYPE_FIXED64,
-	parser.SFixed32: pb.FieldDescriptorProto_TYPE_SFIXED32,
-	parser.SFixed64: pb.FieldDescriptorProto_TYPE_SFIXED64,
-	parser.Bool:     pb.FieldDescriptorProto_TYPE_BOOL,
-	parser.String:   pb.FieldDescriptorProto_TYPE_STRING,
-	parser.Bytes:    pb.FieldDescriptorProto_TYPE_BYTES,
-}
-
-func (b *messageBuilder) createField(pf *parser.Field) *pb.FieldDescriptorProto {
-	if pf.Direct == nil || pf.Direct.Type.Map != nil {
-		panic(fmt.Sprintf("%s: non-direct not implemented", pf.Pos))
+func readProto(file string, importPaths []string, done map[string]bool) ([]*Proto, error) {
+	if done[file] {
+		return nil, nil
 	}
-
-	tag := int32(pf.Direct.Tag)
-	df := &pb.FieldDescriptorProto{
-		Name:           &pf.Direct.Name,
-		Number:         &tag,
-		JsonName:       jsonStr(pf.Direct.Name),
-		Label:          fieldLabel(pf, b.proto3),
-		Proto3Optional: proto3Optional(pf, b.proto3),
+	done[file] = true
+	p, err := newProtoFromPath(file, importPaths)
+	if err != nil {
+		return nil, err
 	}
-
-	if pf.Direct.Type.Reference != nil {
-		t := pb.FieldDescriptorProto_TYPE_MESSAGE
-		df.Type = &t
-
-		name := b.types.fullName(*pf.Direct.Type.Reference, b.scope)
-		df.TypeName = &name
-		return df
+	protos, err := readProtos(p.imports, importPaths, done)
+	if err != nil {
+		return nil, err
 	}
-
-	fieldType := scalars[pf.Direct.Type.Scalar]
-	df.Type = &fieldType
-	return df
-}
-
-func fieldLabel(pf *parser.Field, proto3 bool) *pb.FieldDescriptorProto_Label {
-	var label pb.FieldDescriptorProto_Label
-	switch {
-	case pf.Required:
-		label = pb.FieldDescriptorProto_LABEL_REQUIRED
-	case pf.Repeated:
-		label = pb.FieldDescriptorProto_LABEL_REPEATED
-	case pf.Optional:
-		label = pb.FieldDescriptorProto_LABEL_OPTIONAL
-	case proto3:
-		// unlabelled proto3 field
-		label = pb.FieldDescriptorProto_LABEL_OPTIONAL
-	default:
-		panic(fmt.Sprintf("%s: invalid field label for syntax", pf.Pos))
-	}
-	return &label
+	protos = append(protos, p)
+	return protos, nil
 }
 
-func proto3Optional(pf *parser.Field, proto3 bool) *bool {
-	if pf.Optional && proto3 {
-		p3Optional := true
-		return &p3Optional
+func search(file string, importPaths []string) (io.ReadCloser, error) {
+	if file == "-" {
```

Comment:
Is this a good idea to have this deep in the code? I would expect something like this to be handled in the CLI interface.

Reply:
yeah hmm. probably not a great idea. `protoc -o all.pb -` doesn't even work. removed.